File: engine/src/core/HeirarchyNode.cpp
//
//  HeirarchyNode.cpp
//  mythreal
//
//  Created by Ian Ewell on 7/28/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#include <iostream>
#include "HeirarchyNode.h"
#include "Resource.h"
#include "3DModelResource.h"
#include "LevelLayoutResource.h"

CCHeirarchyNode::CCHeirarchyNode()
{
    mParent = NULL;
    
    mName = "";
    mType = "Generic";
    
    mDepth = 0;
    mID = 0;
}

CCHeirarchyNode::~CCHeirarchyNode()
{
    
    for (std::vector<CCHeirarchyNode*>::iterator it = mChildren.begin(); it < mChildren.end(); it++)
    {
        delete (*it);
    }
}

void CCHeirarchyNode::insertChild(CCHeirarchyNode *child)
{
    child->setID(mChildren.size());
    child->setDepth(mDepth+1);
    child->setParent(this);
    mChildren.push_back(child);
}

CCHeirarchyNode* CCHeirarchyNode::getChildAtIndex(U32 index)
{
    return mChildren.at(index);
}

std::vector<CCHeirarchyNode*> CCHeirarchyNode::getChildren()
{
    return mChildren;
}

CCHeirarchyNode* CCHeirarchyNode::findChildWithName(const std::string &name, bool resursive)
{
    for (std::vector<CCHeirarchyNode*>::iterator it = mChildren.begin(); it < mChildren.end(); it++)
    {
        if ((*it)->getName().compare(name) == 0)
            return (*it);
        else if (resursive && (*it)->getNumberOfChildren() > 0)
        {
            CCHeirarchyNode *node2;
            if ((node2 = (*it)->findChildWithName(name, true)))
                return node2;
        }
    }
    return NULL;
}

U32 CCHeirarchyNode::getNumberOfChildren()
{
    return mChildren.size();
}

std::string CCHeirarchyNode::getNodeType()
{
    return mType;
}

void CCHeirarchyNode::setParent(CCHeirarchyNode *parent)
{
    mParent = parent;
}

CCHeirarchyNode* CCHeirarchyNode::getParent()
{
    return mParent;
}

void CCHeirarchyNode::removeChildAtIndex(U32 index)
{
    mChildren.erase(mChildren.begin()+index);
}

void CCHeirarchyNode::removeSelf()
{
    if (mParent)
        mParent->removeChildAtIndex(mID);
}

void CCHeirarchyNode::readFromStream(CCFileStream *stream)
{
    readChildrenFromStream(stream);
}

void CCHeirarchyNode::readChildrenFromStream(CCFileStream *stream)
{
    U32 childrenCount = stream->readUInt();
    for (int i = 0; i < childrenCount; i++)
    {
        CCHeirarchyNode *child;
        std::string name = stream->readString();
        std::string type = stream->readString();
        
        if (type.compare("Resource") == 0)
            child = new CRResourceNode;
        else if (type.compare("MaterialGroup") == 0)
            child = new CR3DModelMaterialGroup;
        else if (type.compare("3DModelSkeletonNode") == 0)
            child = new CR3DModelSkeletonNode;
        else if (type.compare("3DModelAnimation") == 0)
            child = new CR3DModelAnimation;
        else if (type.compare("2DPhysicsLine") == 0)
            child = new CR2DPhysicsLine;
        else if (type.compare("2DArea") == 0)
            child = new CRArea;
        else if (type.compare("SceneLight") == 0)
            child = new CRLight;
        else if (type.compare("3DSceneObjectInstance") == 0)
            child = new CRSceneObject;
        else
            child = new CCHeirarchyNode;
        
        child->setName(name);
        insertChild(child);
        child->readFromStream(stream);
    }
}

File: editor/src/models/HeirarchyNode.m
/*
 *
 * FILE: HeirarchyNode.m
 *
 * AUTHOR: Ian Ewell
 *
 * PROJECT: Mythreal Engine Resource Editor
 *
 * MODULE: Models
 *
 * PURPOSE: Class for a single hierarchy node.
 *
 * CREATED ON: 07/17/10
 *
 * COPYRIGHT: (C) Copyright 2010 Mythreal Studios. All rights reserved.
 *
 */

#import "HeirarchyNode.h"
#import "Resource.h"
#import "ResourceListModel.h"
#import "3DModelResource.h"
#import "2DSceneObjects.h"
#import "3DSceneObjectInstance.h"

@implementation CHeirarchyNode

- (id)init;
{
	if (self = [super init])
	{
        name = [[NSString alloc] init];
        mChildren = [[NSMutableArray alloc] initWithCapacity:5];
	
        depth = 0;
        ID = 0;
	
        if (!mNodeType)
            mNodeType = [[NSString alloc] initWithString:@"Generic"];
    }
	
	return self;
}

- (void)dealloc
{
	[name release];
	[mChildren release];
	[mNodeType release];
	[super dealloc];
}

- (void)insertChild:(CHeirarchyNode*) child
{
	[child setID:[mChildren count]];
	[child setDepth: [self depth] + 1];
	[child setParent:self];
	[child retain];
	[mChildren addObject:child];
}

- (CHeirarchyNode*)getChildAtIndex:(NSUInteger)index
{
	return [mChildren objectAtIndex:index];
}

- (NSMutableArray*)getChildren
{
    return mChildren;
}

- (CHeirarchyNode*)findChildWithName:(NSString*)nodename isRecursive:(BOOL)recursive
{
    for (CHeirarchyNode *node in mChildren)
    {
        if ([[node name] compare:nodename] == NSOrderedSame)
            return node;
        else if (recursive && [node numberOfChildren] > 0)
        {
            CHeirarchyNode* node2;
            if ((node2 = [node findChildWithName:nodename isRecursive:TRUE]))
                return node2;
        }
    }
    return nil;
}

- (NSUInteger)numberOfChildren
{
	return [mChildren count];
}

- (NSString*)getNodeType
{
	return mNodeType;
}

- (void)setParent: (CHeirarchyNode*)parent
{
	mParent = parent;
}

- (CHeirarchyNode*)getParent
{
    return mParent;
}

- (void)removeChildAtIndex:(NSUInteger)index
{	
	NSUInteger i;
	for (i = index; i < [mChildren count]; i++)
		[[self getChildAtIndex:i] setID:[[self getChildAtIndex:i] ID] - 1];
    [mChildren removeObjectAtIndex:index];
}

- (void)removeSelf
{
	if (mParent)
	{
		[mParent removeChildAtIndex:ID];
	}
	
	[self release];
}

- (void)serializeXML:(NSXMLElement*)parent bundleLocation:(NSString*) loc
{
    NSXMLElement *XMLnode = [[NSXMLElement alloc] initWithName:@"node"];
    
    NSXMLNode *nodeName = [NSXMLNode elementWithName:@"name" stringValue:name];
    NSXMLNode *nodeType = [NSXMLNode elementWithName:@"type" stringValue:mNodeType];
    
    [XMLnode addChild:nodeName];
    [XMLnode addChild:nodeType];
    
    for (CHeirarchyNode *node in mChildren)
        [node serializeXML:XMLnode bundleLocation:loc];
    
    [parent addChild:XMLnode];
    [XMLnode release];
}

- (void)writeToStream:(CFileStream*)stream
{
    // Write our name and type
    [stream writeString:name];
    [stream writeString:mNodeType];
    
    // Write the number of children and the children
    [stream writeUInt:[mChildren count]];
    
    for (CHeirarchyNode *node in mChildren)
        [node writeToStream:stream];
}

- (void)readFromStream:(CFileInputStream*)stream
{
    //name = [stream readString];
    //mNodeType = [stream readString];
    
    // Get the number of children
    unsigned int childrenCount = [stream readUInt];
    int i;
    for (i = 0; i < childrenCount; i++)
    {
        NSString *nodeName = [stream readString];
        NSString *nodeType = [stream readString];
        CHeirarchyNode *child;
        
        if ([nodeType compare:@"Resource"] == NSOrderedSame)
            child = [[CResourceNode alloc] init];
        else if ([nodeType compare:@"ResourceType"] == NSOrderedSame)
            child = [[CResourceType alloc] init];
        else if ([nodeType compare:@"3DModelRoot"] == NSOrderedSame)
            child = [[CModelRoot alloc] init];
        else if ([nodeType compare:@"MaterialGroup"] == NSOrderedSame)
            child = [[CModelMaterialGroup alloc] init];
        else if ([nodeType compare:@"2DPhysicsLine"] == NSOrderedSame)
            child = [[C2DPhysicsLine alloc] init];
        else if ([nodeType compare:@"2DArea"] == NSOrderedSame)
            child = [[C2DArea alloc] init];
        else if ([nodeType compare:@"3DSceneObjectInstance"] == NSOrderedSame)
            child = [[C3DSceneObjectInstance alloc] init];
        else if ([nodeType compare:@"3DModelSkeletonNode"] == NSOrderedSame)
            child = [[CModelSkeletonNode alloc] init];
        else if ([nodeType compare:@"3DModelAnimation"] == NSOrderedSame)
            child = [[CModelAnimation alloc] init];
        else
            child = [[CHeirarchyNode alloc] init];
        
        [child setName:nodeName];
        [self insertChild:child];
        [child readFromStream:stream];
    }
}

@synthesize name;
@synthesize ID;
@synthesize depth;

@end
